alea/runner.py

```diff
@@ -0,0 +1,254 @@
+                result_names[i] = ea
+
+        metadata['date'] = datetime.now().strftime('%Y%m%d_%H:%M:%S')
+        metadata['poi'] = self.poi
+        metadata['common_hypothesis'] = self.common_hypothesis
+        metadata['generate_values'] = self.generate_values
+
+        array_metadatas = [{'hypotheses_values': ea} for ea in self._hypotheses_values]
+        numpy_arrays_and_names = [(r, rn) for r, rn in zip(results, result_names)]
+
+        print(f'Saving {self._output_file}')
+        numpy_to_toyfile(
+            self._output_file,
+            numpy_arrays_and_names=numpy_arrays_and_names,
+            metadata=metadata,
+            array_metadatas=array_metadatas)
+
+    def read_toydata(self):
+        """Read toydata from file"""
+        toydata, toydata_names = toydata_from_file(self._toydata_file)
+        return toydata, toydata_names
+
+    def write_toydata(self, toydata, toydata_names):
+        """
+        Write toydata to file.
+        If toydata is a list of dict, convert it to a list of list.
+        """
+        self.statistical_model.store_data(self._toydata_file, toydata, toydata_names)
+
+    def toy_simulation(self):
+        """
+        For each Monte Carlo:
+            - run toy simulation a specified toydata mode and generate values.
+            - loop over hypotheses.
+        """
+        if self._toydata_mode not in {
+            'read', 'generate', 'generate_and_write', 'no_toydata',
+        }:
+            raise ValueError(f'Unknown toydata mode: {self._toydata_mode}')
+
+        toydata = []
+        toydata_names = None
+        if self._toydata_mode == 'read':
+            toydata, toydata_names = self.read_toydata()
+
+        results = [np.zeros(self._n_mc, dtype=self._result_dtype) for _ in self._hypotheses_values]
+        for i_mc in tqdm(range(self._n_mc)):
+            if self._toydata_mode == 'generate' or self._toydata_mode == 'generate_and_write':
+                self.statistical_model.data = self.statistical_model.generate_data(
+                    **self.generate_values)
+            fit_results = []
+            for hypothesis_values in self._hypotheses_values:
+                if self._toydata_mode == 'read':
+                    self.statistical_model.data = toydata[i_mc]
+
+                fit_result, max_llh = self.statistical_model.fit(**hypothesis_values)
+                fit_result['ll'] = max_llh
+                if self._compute_confidence_interval and (self.poi not in hypothesis_values):
+                    dl, ul = self.statistical_model.confidence_interval(
+                        poi_name=self.poi,
+                        best_fit_args=self._hypotheses_values[0],
+                        confidence_interval_args=hypothesis_values)
+                else:
+                    dl, ul = np.nan, np.nan
+                fit_result['dl'] = dl
+                fit_result['ul'] = ul
+
+                fit_results.append(fit_result)
+            # appemd toydata
```

Comment:
to reduce RAM usage, I think we should only store this if we're planning to write it out

Reply:
updated.